sdk/typescript/src/keypairs/passkey/keypair.ts

```diff
@@ -0,0 +1,223 @@
+import { blake2b } from '@noble/hashes/blake2b';
+import { randomBytes } from '@noble/hashes/utils';
+import type {
+	AuthenticationCredential,
+	RegistrationCredential,
+} from '@simplewebauthn/typescript-types';
+
+import { PasskeyAuthenticator } from '../../bcs/bcs.js';
+import type { IntentScope, SignatureWithBytes } from '../../cryptography/index.js';
+import { messageWithIntent, SIGNATURE_SCHEME_TO_FLAG, Signer } from '../../cryptography/index.js';
+import type { PublicKey } from '../../cryptography/publickey.js';
+import type { SignatureScheme } from '../../cryptography/signature-scheme.js';
+import {
+	parseDerSPKI,
+	PASSKEY_PUBLIC_KEY_SIZE,
+	PASSKEY_SIGNATURE_SIZE,
+	PasskeyPublicKey,
+} from './publickey.js';
+
+export interface PasskeyCreateOptions {
+	/** The timeout for the passkey creation dialog. */
+	timeout: number;
+	/** The name of the user. */
+	name: string;
+	/** The display name of the user, this is shown at creation dialog. */
+	displayName: string;
+	/** The name of the relying party, this is shown to user in the passkey creation dialog. */
+	rpName: string;
+}
+type DeepPartialConfigKeys = 'rp' | 'user' | 'authenticatorSelection';
+
+type DeepPartial<T> = T extends object
+	? {
+			[P in keyof T]?: DeepPartial<T[P]>;
+		}
+	: T;
+
+type BrowserPasswordProviderOptions = Pick<
+	DeepPartial<PublicKeyCredentialCreationOptions>,
+	DeepPartialConfigKeys
+> &
+	Omit<
+		Partial<PublicKeyCredentialCreationOptions>,
+		DeepPartialConfigKeys | 'pubKeyCredParams' | 'challenge'
+	>;
+
+export interface PasskeyProvider {
+	create(): Promise<RegistrationCredential>;
+	get(challenge: Uint8Array): Promise<AuthenticationCredential>;
+}
+
+// Default browser implementation
+export class BrowserPasskeyProvider implements PasskeyProvider {
+	#name: string;
+	#options: BrowserPasswordProviderOptions;
+
+	constructor(name: string, options: BrowserPasswordProviderOptions) {
+		this.#name = name;
+		this.#options = options;
+	}
+
+	async create(): Promise<RegistrationCredential> {
+		return (await navigator.credentials.create({
+			publicKey: {
+				timeout: 60000,
+				...this.#options,
+				rp: {
+					name: this.#name,
+					...this.#options.rp,
+				},
+				user: {
+					name: this.#name,
+					displayName: this.#name,
+					...this.#options.user,
+					id: randomBytes(10),
+				},
+				challenge: new TextEncoder().encode('Create passkey wallet on Sui'),
+				pubKeyCredParams: [{ alg: -7, type: 'public-key' }],
+				authenticatorSelection: {
+					...this.#options.authenticatorSelection,
+					authenticatorAttachment: 'cross-platform',
+					residentKey: 'required',
+					requireResidentKey: true,
+					userVerification: 'required',
+				},
+			},
+		})) as RegistrationCredential;
+	}
+
+	async get(challenge: Uint8Array): Promise<AuthenticationCredential> {
+		return (await navigator.credentials.get({
+			publicKey: {
+				challenge,
+				userVerification: 'required',
+				timeout: this.#options.timeout,
```

Comment:
is this optional in the underlying type, and if so, do we want a default?

Reply:
if its a required option already, ignore this comment!

Reply:
with deepPartial, I think everything is optional, so we probably do need a default here

Reply:
added!